main.go

```diff
@@ -126,7 +133,7 @@ func run(ctx context.Context, env Environment) error {
 	// demonstrates database authentication with dynamic secrets
 	r.GET("/products", h.GetProducts)
 
-	r.Run(env.MyAddress)
+	endless.ListenAndServe(env.MyAddress, r)
```

Comment:
I'm feeling really hesitant on this one... while it probably adds some useful functionality, I think many devs will be unfamiliar with it and it kind of adds another level of "wait what's that, what's going on", as opposed to the familiar-looking r.Run.

I think our application is meant to give an opinionated example of Vault best practices, not opinions on general web app productionizing. 

WaitGroups are good here, since they'd stdlib and quite commonly used, but you can see even Gin seems to not be using endless in their own Run function, they just link to it from their readme since someone asked about it. (Which we don't need to do, since "how to make gin's router more production-ready" is gin's concern, not ours.) Hope that makes sense.

Reply:
I believe the reason why Gin is not doing it on their end is because `gin-gonic` is more of a request-routing framework/library. They provide you with a router which can then be plugged into the application. It would be a poor design choice for gin to integrate too much functionality into the library. Instead, it gives you flexibility to pick whichever ListenAndServer behavior you want (default/manual implementation/3rd party lib).

> I think many devs will be unfamiliar with it and it kind of adds another level of "wait what's that, what's going on", as opposed to the familiar-looking r.Run.

I don't think `r.Run()` is standard, it's just a convenient wrapper. `http.ListenAndServer(address, r)` is much more common from what I've seen. `endless.ListenAndServe(env.MyAddress, r)` simply replaces `http` with `endless` and follows the same familiar pattern. I think it would introduce less of a "surprise factor" to new developers when compared with `r.Run()`. I can also add a comment on top, something along the lines of "http.ListenAndServe with graceful shutdown"

> WaitGroups are good here, since they'd stdlib and quite commonly used

Unfortunately, without a graceful shutdown, the app will simply exit on `SIGTERM` from docker stop. If a developer were to run it and look for the shutdown logs, they will not find them and wonder whether the go-routine logic is correct.

Reply:
I see, you're right that r.Run() is not standard; I think I just see the r and don't get scared because I figure it's either a wrapper we defined or something defined by gin, whereas the word "endless" reveals the existence of some other less well-known library and thus scares and overwhelms me. 😛 But maybe I'm being too minimalist.

Adding the comment `# http.ListenAndServe with graceful shutdown` above the line might be enough to solve my issue with it. What do you think, @VinnyHC ?

I think I'm also a little confused about what you mean by the shutdown logs thing though. So far, while developing the app, I feel like I had plenty of problems and was still able to just go look at the container logs from the moment before the app crashed to see what went wrong. I was using the docker desktop UI, but I'm pretty sure that's just doing `docker ps -a` to see all stopped containers and then doing `docker logs` on that container ID... let me know if I'm missing something.

Reply:
> But maybe I'm being too minimalist.

I also like the code to be minimalistic. I was briefly considering implementing the graceful shutdown logic myself, but then it would actually add a lot more code bloat, which is not related to the main purpose of this project -- showcasing the vault behavior. Using a third party library seemed more minimalist too me 😄 

> I think I'm also a little confused about what you mean by the shutdown logs thing though. 

Sorry for not being clear, what I meant is that without some kind of graceful shutdown mechanism, `http.ListenAndServe` (or the `r.Run` wrapper) will simply exit on SIGTERM (the function never returns back to the caller), so our go routine will never be cleanly shut down and the logs will abruptly stop.

Reply:
I could go either way here (3rd party vs hand-rolled) but I agree that the graceful shutdown is a nice addition. I think the 3rd party library is a good fit here to cut down on bloat. Typically the apps I've worked on in the past do something like this to clean up:
```go
	go func() {
		if err := srv.ListenAndServe(); err != nil {
			log.Printf("shutting down the server: %s", err)
		}
	}()

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	<-c
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
	defer cancel()
	_ = srv.Shutdown(ctx)
	os.Exit(0)
```
But that library seems to be a lot more thorough.

Reply:
Yeap, I've done something similar as well with a bit more logic to handle shutdown errors